File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {

    public class Node {
        int key;
        int value;
        Node next;
        Node prev;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    Node head, tail;
    int capacity;
    HashMap<Integer, Node> keyNode;

    public LRUCache(int capacity) {
        head = new Node(-1, -1);
        tail = new Node(-1, -1);
        keyNode = new HashMap<Integer, Node>();
        head.next = tail;
        tail.prev = head;
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if(!keyNode.containsKey(key)) {
            return -1;
        }
        Node node = keyNode.get(key);
        removeNode(node);
        add(node);
        return node.value;
    }

    private void removeNode(Node node) {
        node.next.prev = node.prev;
        node.prev.next = node.next;
    }
    
    public void put(int key, int value) {
        if(keyNode.containsKey(key)) {
            removeNode(keyNode.get(key));
        }
        Node node = new Node(key, value);
        add(node);
        keyNode.put(key, node);
        if(keyNode.size() > capacity) {
            Node nodeToBeDeleted = head.next;
            removeNode(nodeToBeDeleted);
            keyNode.remove(nodeToBeDeleted.key);
        }
    }

    private void add(Node node) {
        Node prevTail = tail.prev;
        node.prev = prevTail;
        node.next = tail;
        tail.prev = node;
        prevTail.next = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.cs
public class LRUCache {

    public class ListNode {
        public int key;
        public int val;
        public ListNode next;
        public ListNode prev;

        public ListNode(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }

    private ListNode tail = new ListNode(-1, -1);
    private ListNode head = new ListNode(-1, -1);

    private int cap;

    private Dictionary<int, ListNode> map = new Dictionary<int, ListNode>();

    public LRUCache(int capacity) {
        cap = capacity;
        head.next = tail;
        tail.prev = head;
    }
    
    public int Get(int key) {
        if(map.ContainsKey(key)) {
            ListNode node = map[key];
            RemoveNode(node);
            AddNode(node);
            return node.val;
        }
        return -1;
    }

    private void RemoveNode(ListNode node) {
        node.next.prev = node.prev;
        node.prev.next = node.next;
    }
    
    public void Put(int key, int value) {
        if(map.ContainsKey(key)) {
            ListNode nodeToBeDeleted = map[key];
            RemoveNode(nodeToBeDeleted);
            map.Remove(key);
        }
        ListNode node = new ListNode(key, value);
        AddNode(node);
        map[key] = node;
        if(map.Count > cap) {
            ListNode lru = head.next;
            RemoveNode(lru);
            map.Remove(lru.key);
        }

    }

    private void AddNode(ListNode node) {
        // ListNode node = new ListNode(key, value);
        ListNode prevTail = tail.prev;
        node.prev = prevTail;
        prevTail.next = node;
        node.next = tail;
        tail.prev = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */